src/psbt/finalizer.rs

```diff
@@ -347,29 +359,46 @@ pub fn finalize_helper<C: secp256k1::Verification>(
 
     // Actually construct the witnesses
     for index in 0..psbt.inputs.len() {
-        let (witness, script_sig) = {
-            let spk = get_scriptpubkey(psbt, index).map_err(|e| Error::InputError(e, index))?;
-            let sat = PsbtInputSatisfier::new(&psbt, index);
+        finalize_input(psbt, index, secp, allow_mall)?;
+    }
+    // Double check everything with the interpreter
+    // This only checks whether the script will be executed
+    // correctly by the bitcoin interpreter under the current
+    // psbt context.
+    interpreter_check(&psbt, secp)?;
+    Ok(())
+}
 
-            if spk.is_v1_p2tr() {
-                // Deal with tr case separately, unfortunately we cannot infer the full descriptor for Tr
-                let wit = construct_tap_witness(spk, &sat, allow_mall)
-                    .map_err(|e| Error::InputError(e, index))?;
-                (wit, Script::new())
-            } else {
-                // Get a descriptor for this input.
-                let desc = get_descriptor(&psbt, index).map_err(|e| Error::InputError(e, index))?;
+pub(super) fn finalize_input<C: secp256k1::Verification>(
+    psbt: &mut Psbt,
+    index: usize,
+    secp: &Secp256k1<C>,
+    allow_mall: bool,
+) -> Result<(), super::Error> {
+    let (witness, script_sig) = {
+        let spk = get_scriptpubkey(psbt, index).map_err(|e| Error::InputError(e, index))?;
+        let sat = PsbtInputSatisfier::new(&psbt, index);
```

Comment:
In 815d012f2280e89d609c57c2d7357bd86b69c545:

I think we should make this function generic over the satisfier type, and expose additional methods in the `PsbtExt` trait that users can use to provide their own `Satisfier`s. (The existing methods would have this line `let sat = PsbtInputSatisfier::new(&psbt, index);` and pass that into this function.)

I think this, maybe in conjunction with some extensions to the `Satisfier` trait, would give Jeremy the tools to implement what he's getting at here.

Reply:
i actually think that here, this is OK. Since we're finalizing here, not signing, all the data should already be in here.

What might work is an API that takes a PSBT input as a Processing(PsbtInputRef) and then you can call many `fn try_sat(self, s: dyn Satisfier) -> Processing` and then lastly call `Processing::finalize_input(self) -> Result<Psbt, Error>` or something like that which gives you back your Psbt in the end.

This way you could try passing the input in the processing phase to multiple signers, without doing intepreter checks, and then do a final "try to finalize call" which then either succeeds or fails.

One important cleanup: if it does fail, we should make sure the PSBT is not modified at all so that we can retry signing.

Reply:
Ah, yep, I understand.

Your proposed API is interesting but I think too complicated for this API (and this major rev).